Sources/ProcedureKit/ProcedureEventQueue.swift

```diff
@@ -0,0 +1,248 @@
+    ///   - block: the block that is asynchronously submitted to the otherQueue
+    internal func dispatchSynchronizedBlock(onOtherQueue otherQueue: DispatchQueueProtocol, block: @escaping () -> Void) {
+        debugAssertIsOnQueue()
+        assert(otherQueue !== self, "Cannot dispatch synchronized block onOtherQueue if otherQueue is self.")
+        
+        // acquire the QoS class of the current block on the receiver queue
+        let originalQueueQoS = DispatchQoS(qosClass: DispatchQueue.currentQoSClass, relativePriority: 0)
+
+        // suspend the current queue so no additional scheduled async blocks will be executed
+        // after the calling code returns (preventing blocked threads)
+        queue.suspend()
+
+        // dispatch async to the other queue
+        //
+        // the original queue, since it is also synchronized with the block, will not schedule
+        // any additional blocks to be executed after the current block returns (i.e. the block
+        // that called dispatchSynchronizedBlock(onOtherQueue:))
+        //
+        // the caller should ideally return quickly after this function returns so that the current 
+        // eventQueueLock is released and can be acquired in the otherQueue
+        //
+        otherQueue.asyncDispatch(minimumQoS: originalQueueQoS) { [originalQueue = self] in
+
+            // on the other queue, aquire the event lock from the original queue
+            // to ensure that all blocks on the original queue are done executing
+
+            originalQueue.eventQueueLock.withCriticalScope {
+                // i.e. this block should be synchronized with *both* queues
+                block()
+            }
+
+            // after the block is complete, resume the original queue
+            originalQueue.queue.resume()
+        }
+    }
+}
+
+// MARK: - QueueProvider
+
+/**
+ A QueueProvider provides a queue conforming to DispatchQueueProtocol.
+ 
+ Dispatch.DispatchQueue and ProcedureKit.EventQueue both provide themselves.
+ ProcedureKit.Procedure provides its EventQueue.
+ 
+ Several methods in ProcedureKit can take a QueueProvider. For example,
+ ProcedureProtocol's `add*BlockObserver(synchronizedWith:block:)` methods take
+ a QueueProvider for the `synchronizedWith` parameter. This can be a
+ DispatchQueue, a ProcedureKit.EventQueue, or a Procedure.
+ 
+ */
+public protocol QueueProvider {
+    var providedQueue: DispatchQueueProtocol { get }
+}
+
+extension DispatchQueue: QueueProvider {
+    public var providedQueue: DispatchQueueProtocol { return self }
+}
+
+extension EventQueue: QueueProvider {
+    public var providedQueue: DispatchQueueProtocol { return self }
+}
+
+extension Procedure: QueueProvider {
+    public var providedQueue: DispatchQueueProtocol { return eventQueue }
+}
+
+// MARK: - DispatchQueueProtocol
+
+public protocol DispatchQueueProtocol: class {
+    @discardableResult func asyncDispatch(block: @escaping () -> Void) -> DispatchWorkItem
```

Comment:
I think it would be a bit cleaner to drop the `Dispatch` from these functions, if possible. That might cause collisions with the `DispatchQueue` implementation but I don't think so. Just makes it a bit easier to read, for the same reasons they renamed the `DispatchQueue` functions in Swift 3.